.circleci/config.yml

```diff
@@ -2062,16 +2064,101 @@ jobs:
       - run:
           name: Build kona host
           command: just build-kona-host
-          working_directory: kona
+          working_directory: kona-proofs
       - save_cache:
-          key: kona-host-{{ checksum "./kona/justfile" }}-{{ checksum "./kona/version.json" }}
+          key: kona-host-{{ checksum "./kona-proofs/justfile" }}-{{ checksum "./kona-proofs/version.json" }}
           name: Save Kona host to cache
           paths:
-            - "kona/bin/kona-host"
+            - "kona-proofs/bin/kona-host"
       - persist_to_workspace:
           root: .
           paths:
-            - "kona/bin/kona-host"
+            - "kona-proofs/bin/kona-host"
+
+  # Build a single Rust binary for sysgo tests (parameterized).
+  rust-binary-build:
+    docker:
+      - image: <<pipeline.parameters.default_docker_image>>
+    resource_class: xlarge
+    parameters:
+      cache_name:
+        description: "Identifier for cache key and logging"
+        type: string
+      directory:
+        description: "Directory containing the Cargo workspace"
+        type: string
+      build_command:
+        description: "Cargo build command to run"
+        type: string
+      check_binary:
+        description: "Binary name to check for cache hit (in target/release/)"
+        type: string
+      needs_clang:
+        description: "Whether to install clang (needed by bindgen for reth-mdbx-sys)"
+        type: boolean
+        default: false
+    steps:
+      - utils/checkout-with-mise:
+          checkout-method: blobless
+      - run:
+          name: Initialize << parameters.directory >> submodule (shallow)
+          command: git submodule update --init --recursive --depth 1 << parameters.directory >>
+      - restore_cache:
+          name: Restore << parameters.cache_name >> cache
+          keys:
+            - rust-<< parameters.cache_name >>-v2-{{ checksum "<< parameters.directory >>/Cargo.lock" }}
+            - rust-<< parameters.cache_name >>-v2-
```

Comment:
We're only caching the `target/` subdirectory, right? It only contains build outputs, so I'm not sure there's any benefit to restoring any other cache than the exact cache key.

Restoring a previous non-exact-hit cache only makes sense if we were caching build dependencies, which we currently aren't. We may want to do this _using a separate cache_ if it turns out that we're often changing the submodule.

Reply:
caching `<< parameters.directory >>/target` includes compiled dependencies (`target/release/deps/`), build artifacts, and the final binary.

So the fallback key does have value. If only your source changed but dependencies didn't, Cargo can reuse the cached dependency artifacts and just recompile your crate.